### PR TITLE
Faster link highlighting with more types supported

### PR DESCRIPTION
**Pull request by tim-gromeyer**
1. Faster link highlighting
2. More link types supported
3. Don't highlight HTML tags with a `.` in it (not true anymore, since it wes requested to support files...)

### CODE REVIEW

File: markdownhighlighter.cpp [hunk 1]
```diff
@@ -2051,7 +2026,7 @@ void MarkdownHighlighter::highlightInlineRules(const QString &text) {
 */
 int MarkdownHighlighter::highlightInlineSpans(const QString &text,
                                               int currentPos, const QChar c) {
-    //clear code span ranges for this block
+    // clear code span ranges for this block
```
Comment (tim-gromeyer) [inline markdownhighlighter.cpp:2029]: Nothing changed below this spot, except the formatting.  The next change is in the next file

**Comment by tim-gromeyer** [comment]
I noticed something: It highlights HTML tags as links, I need to fix this

**Comment by pbek** [comment]
Thank you for your concern.
You ripped a lot of code out. It's hard to tell what implications that would have.

@Waqar144, opinions?

**Comment by tim-gromeyer** [comment]
> You ripped a lot of code out. It's hard to tell what implications that would have.

 Yeah, I ripped the old code out which used regular expressions. It was just to unreliable so I replaced it

**Comment by tim-gromeyer** [comment]
It is fixed now!

**Comment by tim-gromeyer** [comment]
> You ripped a lot of code out. It's hard to tell what implications that would have.

You can test it by using the following file (https://raw.githubusercontent.com/tim-gromeyer/html2md/main/tests/links.md) to check the highlighting. It should match the behavior as it did before.

You can copy the content of the linked file and try out the old behavior here (https://tim-gromeyer.github.io/MarkdownEdit/markdownedit.html). Then, compare it to a locally built version with this PR integrated.

**Comment by Waqar144** [comment]
> And sorry... my IDE changed a bit of formatting in some lines... but no feature changes

Yes, but it makes reviewing harder than it needs to be as one has to go through changes that are meaningless. Thus I would suggest you to remove the formatting changes, or at least put them in a separate commit, then rebase the actual feature changes on top. The way it currently is, it is hard to see what is going on.

**Comment by pbek** [comment]
I'd also rather like to have whitespace changes in a different PR...

**Comment by tim-gromeyer** [comment]
For the  links enclosed in angle brackets (`<>`) it is on purpose. Before it also highlighted HTML tags with a dot (`.`) in it. For example `<script src="script.js">`.  Just put `file://` before it and it will work

And for the greyed links... I didn't saw it. You couldn't really see it with dark mode

 What highlighter state `_format[HighlighterState]` does it use?

**Comment by tim-gromeyer** [comment]
@pbek, can you test it again, please? It should be fixed now

**Comment by pbek** [comment]
Those still don't work:

```markdown
## PR 188

- <note://QOwnNotes_Todo>
- <link%20destination.md>
- <linkdestination.md>
- <attachments/1672212884.pdf>
- note://Note_2018_06_26T22_11_10
- <../p11.md>
```

**Comment by pbek** [comment]
> Link highlighting doesn't work some times. 

Can you please give some examples when it doesn't work?
And can you please explain why your changes are more robust?

I'm still very unhappy to lose all the documentation on what cases of links the regular expressions cover...

**Comment by tim-gromeyer** [comment]
> Can you please give some examples when it doesn't work?

I subclassed the highlighter and it has Problems highlighting everything defined by regular expressions

> And can you please explain why your changes are more robust?

I don't know, the regular expressions only work the first time or when pasting something inside, but not after editing the line

I'm using following code:
```cpp
void SpellChecker::highlightBlock(const QString &text)
{
    if (markdownhig)
        MarkdownHighlighter::highlightBlock(text);

    if (spellingEnabled && !language.isEmpty())
        checkSpelling(text);
}
```

> I'm still very unhappy to lose all the documentation on what cases of links the regular expressions cover...

I can add you more documentation if you want to, it's pretty simple

**Comment by pbek** [comment]
> imo, Those shouldn't be handled by this library as they aren't really links I think but only something that QON considers as links. As such, QON should highlight these in it's highlighter subclass. The base class should be more generic

This would be fine with me, but then someone would need to do the migration in https://github.com/pbek/QOwnNotes/blob/main/src/helpers/qownnotesmarkdownhighlighter.cpp first before I'd consider merging that PR...

**Comment by tim-gromeyer** [comment]
I already added support for it here

File: markdownhighlighter.cpp [hunk 2]
```diff
@@ -1996,42 +1882,218 @@ int isInLinkRange(int pos, QVector<QPair<int, int>> &range) {
...
+        QLatin1String("http://"),  QLatin1String("https://"),
+        QLatin1String("file://"),  QLatin1String("www."),
+        QLatin1String("ftp://"),   QLatin1String("mailto:"),
+        QLatin1String("tel:"),     QLatin1String("sms:"),
+        QLatin1String("smsto:"),   QLatin1String("data:"),
+        QLatin1String("irc://"),   QLatin1String("gopher://"),
+        QLatin1String("spotify:"), QLatin1String("steam:"),
+        QLatin1String("bitcoin:"), QLatin1String("magnet:"),
+        QLatin1String("ed2k://"),  QLatin1String("news:"),
+        QLatin1String("ssh://"),   QLatin1String("note://")};
```
Comment (tim-gromeyer) [inline markdownhighlighter.cpp:1921]: `note://` is supported here

**Comment by pbek** [comment]
> I already added support for it here

yes, but just the `note://` ones...

**Comment by tim-gromeyer** [comment]
> > I already added support for it here
> 
> yes, but just the `note://` ones...

What else do you want? The files? They caused problems with some HTML tags being highlighted. They are not highlighted on purpose because of this. You can use `file://` or make a list of all file extensions and check if it ends with a known one to avoid false positives with HTML tags.

**Comment by Waqar144** [comment]
> > > I already added support for it here
> > 
> > 
> > yes, but just the `note://` ones...
> 
> What else do you want? The files? They caused problems with some HTML tags being highlighted. They are not highlighted on purpose because of this. You can use `file://` or make a list of all file extensions and check if it ends with a known one to avoid false positives with HTML tags.

I think he means that we still need to be able to support all the kind of links that previous code did. Some links that may not be "generic" enough and are only needed to by QOwnNotes need to be supported there, so someone, must add support for them in QON before this PR can be considered for merging.

**Comment by tim-gromeyer** [comment]
> I think he means that we still need to be able to support all the kind of links that previous code did

It had a tendency to highlight HTML tags, but of course I'll add support for that

**Comment by tim-gromeyer** [comment]
Works now

**Comment by pbek** [comment]
LGTM, thank you!

**Comment by pbek** [comment]
@tim-gromeyer, unfortunately https://github.com/pbek/qmarkdowntextedit/commit/53c4c598 did break something else too:

```markdown

    wget http://google.com

```

`http://google.com` is now highlighted as link, but it shouldn't.

**Comment by pbek** [comment]
I opened an issue for it #199.

**Pull request merged**